lib/3scale_toolbox/commands/copy_command/copy_service.rb

```diff
@@ -0,0 +1,214 @@
+require 'cri'
+require '3scale_toolbox/base_command'
+
+module ThreeScaleToolbox
+  module Commands
+    module CopyCommand
+      module CopyServiceSubcommand
+        extend ThreeScaleToolbox::Command
+        def self.command
+          Cri::Command.define do
+            name        'service'
+            usage       'service [options] <service_id>'
+            summary     '3scale CLI copy service'
+            description '3scale CLI copy service tool'
+
+            flag :h, :help, 'show help for this command' do |_, cmd|
+              puts cmd.help
+              exit 0
+            end
+
+            required  :s, :source, 'Source'
+            required  :d, :destination, 'Destination'
+            required  :t, 'target_system_name', 'Target system name'
+
+            run do |opts, args, _|
+              CopyServiceSubcommand.run opts, args
+            end
+          end
+        end
+
+        def self.run(opts, args)
+          source      = fetch_required_option(opts, :source)
+          destination = fetch_required_option(opts, :destination)
+          system_name = fetch_required_option(opts, :target_system_name)
+          exit_with_message 'error: missing service_id argument' if args.empty?
+          service_id = args[0]
+          copy_service(service_id, source, destination, system_name)
+        end
+
+        def self.exit_with_message(message)
+          puts message
+          exit 1
+        end
+
+        def self.fetch_required_option(options, key)
+          options.fetch(key) { exit_with_message "error: Missing argument #{key}" }
+        end
+
+        def self.compare_hashes(first, second, keys)
+          keys.map{ |key| first.fetch(key) } == keys.map{ |key| second.fetch(key) }
+        end
+
+        def self.provider_key_from_url(url)
+          url[/\w*@/][0..-2]
+        end
+
+        def self.endpoint_from_url(url)
+          url.sub /\w*@/, ''
+        end
+
+        def self.copy_service_params(original, system_name)
+          {
+            name:        original['name'],
+            system_name: system_name,
+            backend_version: original['backend_version'],
+            end_user_registration_required: original['end_user_registration_required']
+          }.reject { |key, value| !value }
+        end
+
+        def self.copy_service(service_id, source, destination, system_name)
```

Comment:
this method is very long and calls for refactoring into smaller parts

Reply:
Agreed, but it is an improvement over existing one file solution :) This PR is supposed to show off the plugin system, not to reimplement those commands in a nice, easy to test way.